add new rule for securing application insights telemetry (#7014)
SSWConsulting/SSW.Rules.Content: rules/appinsights-authentication/rule.md

---
type: rule
archivedreason: 
title: Do you secure your AppInsights Telemetry?
guid: f2155eb7-6ade-4757-b962-a1dc680cdbbf
uri: appinsights-authentication
created: 2023-10-16T15:11:34.0000000Z
authors:
- title: William Liebenberg
  url: https://ssw.com.au/people/william-liebenberg
- title: Alex Rothwell
  url: https://www.ssw.com.au/people/alex-rothwell/
related: []
redirects: []

---

Azure Application Insights is your app's friendly doctor, keeping an eye on it to stay healthy and running smoothly. It uses an `Instrumentation Key` to allow your app to send telemetry data to its resource.

While this `Instrumentation Key` isn't a secret and is often included in client-side code, it's crucial to keep it safe. If it falls into the wrong hands, it could corrupt your data or lead to extra costs.

The key doesn't grant access to your data; for that, you use Entra ID (used to be Azure AD) authentication.

> To preventing unintended access, Local authentication (typically Instrumentation Key or API Keys) can be **disabled** for authenticating data ingestion. Once disabled, a **stronger authentication** method known as Role **Based Access Control (RBAC)** can be used.

Reviewer: This is a great rule, but I can see this rule can be used in the unintended direction.

I think we should clearly outline the trade-off of securing application insights.
Enabling this effectively blocks sending any client-side telemetries to this AI instance - this approach will require us to do an alternative such as:
- create another application insight and then combine them together somehow
- only use SSR frontend
- create a secure ingestion point in backend (I think this is a bad approach)
- secure frontend with user AD Authentication (locks us down to AD)

What do you think? 😄

Author: I agree with everything you listed above. There will be trade offs.

I can add the PROs and CONs for securing server-side telemetry, but I'll dig a bit further to see what we can recommend for client-side telemetry.

Thanks @christoment !

<!--endintro-->

With Role-Based Access Control (RBAC) we can define the specific applications and users that have the ability to send telemetry data, making unauthorized data injection harder.

This method also allows for more accurate Monitoring and Auditing because it allows us to log who accessed our resources and what actions they performed, which is crucial for security.

> In summary, disabling local authentication and relying on Azure AD Role Based Access Control for telemetry ingestion in Azure Application Insights allows for more robust security measures, ensuring that only authorized entities can submit telemetry data.

Reviewer 2: ```suggestion
In summary, disabling local authentication and relying on Entra ID RBAC for telemetry ingestion in Azure Application Insights allows for more robust security measures, ensuring that only authorized entities can submit telemetry data.
```

![Increase Security - Disable Local Authentication](appinsights-disable-local-auth.png)

In short, to secure the telemetry sent from your application to Application Insights, follow these steps:

1. Use System-assigned or User-assigned Managed Identity with AppService
2. Disable Local Authentication to enable RBAC

> 3. Assign `Monitoring Metrics Publisher` role to the AppService

Reviewer 2: ```suggestion
3. Assign the `Monitoring Metrics Publisher` role to the AppService
```

> 3. Assign `Monitoring Metrics Publisher` role to the AppService

Reviewer 2: ```suggestion
3. Assign the `Monitoring Metrics Publisher` role to the AppService
```

2. Add the following Bicep code to assign the `Monitoring Metrics Publisher` role to an Azure AppService:

```bicep
// Get the existing Monitoring Metrics Publisher role definition
resource monitoringMetricsPublisherRoleDefinition 'Microsoft.Authorization/roleDefinitions@2022-04-01' existing = {
  scope: subscription()
  name: '3913510d-42f4-4e42-8a64-420c390055eb'
}

// Assign the Monitoring Metrics Publisher role to the AppService identity
resource monitoringMetricsPublisherRoleAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
  name: guid(appInsights.id, appService.id, monitoringMetricsPublisherRoleDefinition.id)
  scope: appInsights
  properties: {
    principalType: 'ServicePrincipal'
    roleDefinitionId: monitoringMetricsPublisherRoleDefinition.id
    principalId: appService.identity.principalId
  }
}
```

::: good
Figure: Assigning Monitoring Metrics Publisher role to AppService Managed Identity using Bicep
:::

![Figure: AppInsights shows the Monitoring Metrics Publisher role is assigned to AppService](appinsights-assigned-role.png)

## Authenticating with Application Insights

> For ASP.NET Core applications running on Azure AppServices with the `Monitoring Metrics Publisher` role assigned, we can the use a managed identity (System-assigned or User-assigned) to authenticate the application with a particular AppInsights instance.

Reviewer 2: ```suggestion
For ASP.NET Core applications running on Azure AppServices with the `Monitoring Metrics Publisher` role assigned, we can use a managed identity (System-assigned or User-assigned) to authenticate the application with a particular AppInsights instance.
```
